stackslib/src/chainstate/nakamoto/coordinator/mod.rs

```diff
@@ -243,34 +238,151 @@ pub fn get_nakamoto_reward_cycle_info<U: RewardSetProvider>(
     let reward_cycle = burnchain
         .next_reward_cycle(burn_height)
         .expect("FATAL: no reward cycle for burn height");
-    let reward_start_height = burnchain.reward_cycle_to_block_height(reward_cycle);
 
     debug!("Processing reward set for Nakamoto reward cycle";
           "burn_height" => burn_height,
           "reward_cycle" => reward_cycle,
           "reward_cycle_length" => burnchain.pox_constants.reward_cycle_length,
           "prepare_phase_length" => burnchain.pox_constants.prepare_length);
 
+    let Some((rc_info, anchor_block_header)) = load_nakamoto_reward_set(
+        burn_height,
+        sortition_tip,
+        burnchain,
+        chain_state,
+        sort_db,
+        provider,
+    )?
+    else {
+        return Ok(None);
+    };
+
+    let block_id = match anchor_block_header.anchored_header {
+        StacksBlockHeaderTypes::Epoch2(..) => anchor_block_header.index_block_hash(),
+        StacksBlockHeaderTypes::Nakamoto(ref header) => header.block_id(),
+    };
+
+    info!(
+        "Anchor block selected";
+        "cycle" => reward_cycle,
+        "block_id" => %block_id,
+        "consensus_hash" => %anchor_block_header.consensus_hash,
+        "burn_height" => anchor_block_header.burn_header_height,
+        "anchor_chain_tip" => %anchor_block_header.index_block_hash(),
+        "anchor_chain_tip_height" => %anchor_block_header.burn_header_height,
+    );
+
+    return Ok(Some(rc_info));
+}
+
+/// Helper to get the Nakamoto reward set for a given reward cycle.
+/// In all but the first Nakamoto reward cycle, this will load up the stored reward set from the
+/// Nakamoto chain state.  In the first Nakamoto reward cycle, where the reward set is computed
+/// from epoch2 state, the reward set will be loaded from the sortition DB (which is the only place
+/// it will be stored).
+///
+/// Returns Ok(Some((reward set info, PoX anchor block header))) on success
+/// Returns Ok(None) if the reward set is not yet known, but could be known by the time a
+/// subsequent call is made.
+pub fn load_nakamoto_reward_set<U: RewardSetProvider>(
+    burn_height: u64,
```

Comment:
The `burn_height` parameter in these methods is pretty confusing to me. It gets passed around between a bunch of different functions, and its not at all obvious what exactly its being used for. On closer inspection, it seems like it only ever gets used to compute the reward cycle number, but it is translated back and forth from burn height to reward cycle multiple times. In this method, for example, it's used to figure out what the _next_ reward cycle is (so, this method kind of implicitly assumes its being invoked during the prepare phase to load the next reward cycle's information), but before being passed onto `provider.get_reward_set_nakamoto()` its translated to the first block of the reward cycle, subtracted by 2, and gets passed directly to `read_reward_set_nakamoto()` which only ever uses it to invoke 

```rust
        let cycle = burnchain
            .block_height_to_reward_cycle(cycle_start_burn_height)
            .expect("FATAL: no reward cycle for burn height");
```

Which seems like it could lead to an off-by-one error for the reward cycle (i.e., `reward_cycle` in `load_nakamoto_reward_set` would return `n`, but  `read_reward_set_nakamoto` is using cycle `n-1` because of the `subtract(2)`).

I think it'd be much simpler if these methods just used `reward_cycle` as their parameter here, because the burn height is only ever used to calculate the reward cycle, so we can just set that directly.

Reply:
Sure, will do.